core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -0,0 +1,717 @@
+
+convVarFresh :: Converting m => GHC.Var -> m (PCType, PC.Name ())
+convVarFresh v = do
+    t' <- convType $ GHC.varType v
+    n' <- convNameFresh $ GHC.varName v
+    pure (t', n')
+
+lookupTyName :: Scope -> GHC.Name -> Maybe (PC.TyName ())
+lookupTyName (Scope _ tyns) n = Map.lookup n tyns
+
+safeFreshTyName :: MonadQuote m => String -> m (PC.TyName ())
+safeFreshTyName s = PC.TyName <$> safeFreshName s
+
+convTyNameFresh :: MonadQuote m => GHC.Name -> m (PC.TyName ())
+convTyNameFresh n = PC.TyName <$> convNameFresh n
+
+convTyVarFresh :: Converting m => GHC.TyVar -> m (PC.Kind () , PC.TyName ())
+convTyVarFresh v = do
+    k' <- convKind $ GHC.tyVarKind v
+    t' <- convTyNameFresh $ GHC.varName v
+    pure (k', t')
+
+pushName :: GHC.Name -> PC.Name () -> ScopeStack -> ScopeStack
+pushName ghcName n stack = let Scope ns tyns = NE.head stack in Scope (Map.insert ghcName n ns) tyns NE.<| stack
+
+pushTyName :: GHC.Name -> PC.TyName () -> ScopeStack -> ScopeStack
+pushTyName ghcName n stack = let Scope ns tyns = NE.head stack in Scope ns (Map.insert ghcName n tyns) NE.<| stack
+
+-- Types and kinds
+
+convKind :: Converting m => GHC.Kind -> m (PC.Kind ())
+convKind k = case k of
+    -- this is a bit weird because GHC uses 'Type' to represent kinds, so '* -> *' is a 'TyFun'
+    (GHC.isStarKind -> True)              -> pure $ PC.Type ()
+    (GHC.splitFunTy_maybe -> Just (i, o)) -> PC.KindArrow () <$> convKind i <*> convKind o
+    _                                     -> unsupported $ "Kind:" GHC.<+> GHC.ppr k
+
+convType :: Converting m => GHC.Type -> m PCType
+convType t = do
+    -- See Note [Scopes]
+    (_, _, _, stack) <- ask
+    let top = NE.head stack
+    case t of
+        -- in scope type name
+        (GHC.getTyVar_maybe -> Just (lookupTyName top . GHC.varName -> Just name)) -> pure $ PC.TyVar () name
+        (GHC.getTyVar_maybe -> Just v) -> freeVariable $ "Type variable:" GHC.<+> GHC.ppr v
+        (GHC.splitFunTy_maybe -> Just (i, o)) -> PC.TyFun () <$> convType i <*> convType o
+        (GHC.splitTyConApp_maybe -> Just (tc, ts)) -> convTyConApp tc ts
+        (GHC.splitForAllTy_maybe -> Just (tv, tpe)) -> mkTyForall tv (convType tpe)
+        -- I think it's safe to ignore the coercion here
+        (GHC.splitCastTy_maybe -> Just (tpe, _)) -> convType tpe
+        _ -> unsupported $ "Type" GHC.<+> GHC.ppr t
+
+convTyConApp :: (Converting m) => GHC.TyCon -> [GHC.Type] -> m PCType
+convTyConApp tc ts
+    -- this is Int
+    | tc == GHC.intTyCon = pure $ appSize haskellIntSize (PC.TyBuiltin () PC.TyInteger)
+    -- this is Int#, can we do this nicer?
+    | (GHC.getOccString $ GHC.tyConName tc) == "Int#" = pure $ appSize haskellIntSize (PC.TyBuiltin () PC.TyInteger)
+    | otherwise = do
+        tc' <- convTyCon tc
+        args' <- mapM convType ts
+        -- See Note [Iterated abstraction and application]
+        pure $ foldl' (\acc t -> PC.TyApp () acc t) tc' args'
```

Comment:
In general, we will probably want a map from Haskell tycons to PLC tycons or types. The unboxed types in Haskell are a nuisance here.

You wrote a note (above) about nominal (GHC) versus structured (PLC) types. Are we basically removing all newtypes (including their wrapping and unwrapping, which these days are mostly coercions)? (This will probably require further thinking once we look at recursive types.)

Reply:
Yes, the unboxed types are a big nuisance! My handling of them is somewhat ad hoc, trying to match the both the correct constructor applications...

I'm actually ignoring newtypes at the moment, but I indeed intend to treat them exactly like the underlying type. Mostly they can't mean anything else, but I see that they could indicate a wrap/unwrap point if things are recursive. Recursive types are going to be tricky!